File: backend/src/main/java/com/training/cvmanagementbe/service/CvProfileService.java
package com.training.cvmanagementbe.service;

import com.training.cvmanagementbe.dto.request.CvProfileRequest;
import com.training.cvmanagementbe.dto.request.CvProfileRestoreRequest;
import com.training.cvmanagementbe.dto.response.CvProfileResponse;
import com.training.cvmanagementbe.dto.response.EmployeeTeamResponse;
import com.training.cvmanagementbe.dto.response.PagedResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface CvProfileService {

    // Active profiles of one employee, primary first.
    PagedResponse<CvProfileResponse> listByEmployee(UUID employeeId, Pageable pageable);

    // Teams the employee belongs to, used to populate the linked-team picker.
    List<EmployeeTeamResponse> listAssignableTeams(UUID employeeId);

    CvProfileResponse getById(UUID id);

    CvProfileResponse create(UUID employeeId, CvProfileRequest request);

    CvProfileResponse update(UUID id, CvProfileRequest request);

    // Soft delete; also soft-deletes the profile's CVs.
    void delete(UUID id);

    // Promotes one profile and demotes the previous primary in the same transaction.
    CvProfileResponse setPrimary(UUID id);

    /*
     * Guarantees the employee owns at least one profile.
     *
     * - Idempotent: with a profile already present it returns the current primary and creates
     * nothing. Otherwise it creates the first profile from the employee's primary team.
     */
    CvProfileResponse ensureProfileExists(UUID employeeId);

    /*
     * Restores a soft-deleted profile.
     *
     * - Two conditions: the name must still be free among the employee's ACTIVE profiles
     * (otherwise the request must carry newName) and the linked team must still be one of the
     * employee belongs to (otherwise newTeamId). A restored profile always comes back as a
     * secondary profile - it does not reclaim is_primary.
     * - The CVs inside are not restored with it; each one comes back through CvService restore().
     */
    CvProfileResponse restore(UUID id, CvProfileRestoreRequest request);
}

File: frontend/src/app/core/services/cv-profile.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { inject, Injectable } from "@angular/core";
import { environment } from "../../../environments/environment";
import { ApiEndpoint } from "../enums/api-endpoint.enum";
import { CvProfileQuery, CvProfileRequest, CvProfileResponse } from "../dtos/cv-profile.dto";
import { Observable } from "rxjs";
import { PagedResponse } from "../dtos/page.dto";
import { EmployeeTeamResponse } from "../dtos/employee-team.dto";

@Injectable({ providedIn: 'root' })
export class CvProfileService {

    private readonly http = inject(HttpClient);

    private url(endpoint: string): string {
        return `${environment.apiBaseUrl}${endpoint}`;
    }

    private profilesOf(employeeId: string): string {
        return `${ApiEndpoint.Employees}/${employeeId}/cv-profiles`;
    }

    listByEmployee(query: CvProfileQuery): Observable<PagedResponse<CvProfileResponse>> {
        const params = new HttpParams()
            .set('page', query.page)
            .set('size', query.size);

        return this.http.get<PagedResponse<CvProfileResponse>>(
            this.url(this.profilesOf(query.employeeId)), { params }
        );
    }

    listAssignableTeams(employeeId: string): Observable<EmployeeTeamResponse[]> {
        return this.http.get<EmployeeTeamResponse[]>(
            this.url(`${this.profilesOf(employeeId)}/team-options`)
        );
    }

    getById(id: string): Observable<CvProfileResponse> {
        return this.http.get<CvProfileResponse>(this.url(`${ApiEndpoint.Profiles}/${id}`));
    }

    create(employeeId: string, body: CvProfileRequest): Observable<CvProfileResponse> {
        return this.http.post<CvProfileResponse>(this.url(this.profilesOf(employeeId)), body);
    }

    update(id: string, body: CvProfileRequest): Observable<CvProfileResponse> {
        return this.http.put<CvProfileResponse>(this.url(`${ApiEndpoint.Profiles}/${id}`), body);
    }

    delete(id: string): Observable<void> {
        return this.http.delete<void>(this.url(`${ApiEndpoint.Profiles}/${id}`));
    }

    setPrimary(id: string): Observable<CvProfileResponse> {
        return this.http.post<CvProfileResponse>(this.url(`${ApiEndpoint.Profiles}/${id}/set-primary`), {});
    }

    // Idempotent: returns the current primary profile, creating the first one if none exists.
    ensureProfile(employeeId: string): Observable<CvProfileResponse> {
        return this.http.post<CvProfileResponse>(this.url(`${this.profilesOf(employeeId)}/ensure`), {});
    }
}